### ISSUE TITLE
Assign op pattern should not apply in const context.

### ISSUE THREAD

**Issue by Wasabi375** [rust-lang/rust-clippy#14871]
### Summary

Right now only a handful of traits are actually const traits. One of them is `Add`. However `AddAssign` is not one of them. Therefor 
the statement `a = a + some_constant` is not equal to `a += some_constant`.

### Lint Name

assign_op_pattern

### Reproducer

I tried this code:

```rust
impl<T> const NumberConstants for T
where
    T: Number + ~const core::ops::Add,
{
    fn constant(value: usize) -> Self {
        let mut res = Self::ZERO;
        let mut count = 0;
        while count < value {
            res = res + Self::ONE;
            count += 1;
        }
        res
    }
}
<code>
```

I saw this happen:

```
res += Self::ONE;
```

I expected to see this happen:
Clippy should not apply this pattern when in a const context.

### Version

```text
rustc 1.89.0-nightly (bc8215286 2025-05-20)
binary: rustc
commit-hash: bc821528634632b4ff8dee5ac1ea4ad90d1b3eb5
commit-date: 2025-05-20
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by profetia** [comment]
Can you give a more complete reproducer? I am unable to reproduce a case where `AddAssign` is unstable

**Comment by Wasabi375** [comment]
Yeah sorry, I missed one detail in when simplifying the code. I should have tested the simplified version again.

`Number` inherits `AddAssign` which is crucial for the bug.

This should reproduce correctly
```rust
#![feature(const_trait_impl, const_ops)]

use std::ops::{Add, AddAssign};

pub trait Number: Copy + Add<Self, Output = Self> + AddAssign {
    const ZERO: Self;
    const ONE: Self;
}

#[const_trait]
pub trait NumberConstants {
    fn constant(value: usize) -> Self;
}

impl<T> const NumberConstants for T
where
    T: Number + ~const core::ops::Add,
{
    fn constant(value: usize) -> Self {
        let mut res = Self::ZERO;
        let mut count = 0;
        while count < value {
            res = res + Self::ONE;
            count += 1;
        }
        res
    }
}
```
